Feature/oo adaptivity2

drreynolds opened the pull request:
New object-oriented infrastructure for time step adaptivity controllers. These are currently used in ARKODE. In the not-distant future, we can implement additional step/order controllers within this existing infrastructure to support CVODE(S) and IDA(S) (and maybe even ARKODE).

Also:
- [x] Merge https://github.com/sundials-codes/answers/pull/22

balos1 commented on scripts/shared:
```diff
@@ -176,6 +177,13 @@ $tar $tarfile $distrobase/include/nvector/nvector_trilinos.h
 $tar $tarfile $distrobase/include/nvector/trilinos/SundialsTpetraVectorInterface.hpp
 $tar $tarfile $distrobase/include/nvector/trilinos/SundialsTpetraVectorKernels.hpp
 
+$tar $tarfile $distrobase/include/sunadaptcontroller/sunadaptcontroller_expgus.h
```
This should be removed since it no longer exists (this is causing the Jenkins failure).

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_ImExGus.rst:
```diff
@@ -0,0 +1,129 @@
+
+The default values of :math:`k_1` and :math:`k_2` are 0.367 and 0.268,
+respectively, and :math:`p` is the global order of the time
+integration method.  In this estimate, a floor of :math:`\varepsilon > 10^{-10}`
+is enforced to avoid division-by-zero errors.
```
Maybe this and the next paragraph should be combined and this moved to the end since it applies to both controllers. 
```suggestion
integration method.  In this estimate, a floor of :math:`\varepsilon_* > 10^{-10}`
is enforced to avoid division-by-zero errors.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_ImExGus.rst:
```diff
@@ -0,0 +1,129 @@
+
+
+The SUNAdaptController_ImExGus class provides implementations of all operations
+relevant to a `SUN_ADAPTCONTROLLER_H` controller listed in
```
```suggestion
relevant to a ``SUN_ADAPTCONTROLLER_H`` controller listed in
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_ImExGus.rst:
```diff
@@ -0,0 +1,129 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_ImExGus
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_ImExGus
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+The Soderlind implementation of the SUNAdaptController class,
+SUNAdaptController_Soderlind, implements a general structure for temporal
+control proposed by G. Soderlind in :cite:p:`Sod:98`, :cite:p:`Sod:03`
```
```suggestion
control proposed by G. Soderlind in :cite:p:`Sod:98`, :cite:p:`Sod:03`,
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+with default parameter values :math:`k_1 = 1.25`, :math:`k_2 = 0.5`,
+:math:`k_3 = -0.75`, :math:`k_4 = 0.25` and :math:`k_5 = 0.75`, where
+:math:`p` is the global order of the time integration method.  In this estimate,
+a floor of :math:`\varepsilon > 10^{-10}` is enforced to avoid division-by-zero
```
```suggestion
a floor of :math:`\varepsilon_* > 10^{-10}` is enforced to avoid division-by-zero
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+:math:`p` is the global order of the time integration method.  In this estimate,
+a floor of :math:`\varepsilon > 10^{-10}` is enforced to avoid division-by-zero
+errors.  During the first two steps (when :math:`\varepsilon_{n-2}`,
+:math:`\varepsilon_{n-1}`, :math:`h_{n-2}` and :math:`h_{n-2}` may be unavailable),
```
```suggestion
:math:`\varepsilon_{n-1}`, :math:`h_{n-2}`, and :math:`h_{n-2}` may be unavailable),
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+   h' = h_n \varepsilon_n^{-k_1/(p+1)} \varepsilon_{n-1}^{-k_2/(p+1)} \varepsilon_{n-2}^{-k_3/(p+1)} \left(\dfrac{h_n}{h_{n-1}}\right)^{k_4} \left(\dfrac{h_{n-1}}{h_{n-2}}\right)^{k_5}
+
+with default parameter values :math:`k_1 = 1.25`, :math:`k_2 = 0.5`,
+:math:`k_3 = -0.75`, :math:`k_4 = 0.25` and :math:`k_5 = 0.75`, where
```
```suggestion
:math:`k_3 = -0.75`, :math:`k_4 = 0.25`, and :math:`k_5 = 0.75`, where
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+  estimate via :math:`\varepsilon = \text{bias}*\text{dsm}`.
+
+* ``ep, epp`` - storage for the two previous error estimates,
+  :math:`\varepsilon_{n-1}` and :math:`varepsilon_{n-2}`.
```
```suggestion
  :math:`\varepsilon_{n-1}` and :math:`\varepsilon_{n-2}`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+The header file to be included when using this module is
+``sunadaptcontroller/sunadaptcontroller_soderlind.h``.
+
+We note that through appropriate selection of the parameters :math:`k_1 - k_5`,
```
```suggestion
We note that through appropriate selection of the parameters :math:`k_*`,
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+SUNAdaptController_Soderlind) are provided.
+
+The SUNAdaptController_Soderlind class provides implementations of all operations
+relevant to a `SUN_ADAPTCONTROLLER_H` controller listed in
```
```suggestion
relevant to a ``SUN_ADAPTCONTROLLER_H`` controller listed in
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
+   object, set up to replicate a PID controller, and inserts its default parameters
+   (:math:`k_1=0.58`, :math:`k_2=-0.21`, :math:`k_3=0.1`, and :math:`k_4=k_5=0`).
```
```suggestion
   :math:`k_1=0.58`, :math:`k_2=-0.21`, :math:`k_3=0.1`, and :math:`k_4=k_5=0`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
+   object, set up to replicate a PI controller, and inserts its default parameters
+   (:math:`k_1=0.8`, :math:`k_2=-0.31`, and :math:`k_3=k_4=k_5=0`).
```
```suggestion
   :math:`k_1=0.8`, :math:`k_2=-0.31`, and :math:`k_3=k_4=k_5=0`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: int SUNAdaptController_SetParams_PID(SUNAdaptController C, sunrealtype k1, sunrealtype k2, sunrealtype k3)
+
+   This user-callable function provides control over the relevant parameters
+   above, setting :math:`k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
```
```suggestion
   This user-callable function provides control over the relevant parameters
   above for a PID controller, setting :math:`k_4 = k_5 = 0`.  This should be
   called *before* the time integrator is called to evolve the problem.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: int SUNAdaptController_SetParams_PI(SUNAdaptController C, sunrealtype k1, sunrealtype k2)
+
+   This user-callable function provides control over the relevant parameters
+   above, setting :math:`k_3 = k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
```
```suggestion
   This user-callable function provides control over the relevant parameters
   above for a PI controller, setting :math:`k_3 = k_4 = k_5 = 0`.  This should
   be called *before* the time integrator is called to evolve the problem.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
+   object, set up to replicate an I controller, and inserts its default parameters
+   (:math:`k_1=1.0` and :math:`k_2=k_3=k_4=k_5=0`).
```
```suggestion
   :math:`k_1=1.0` and :math:`k_2=k_3=k_4=k_5=0`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
+   object, set up to replicate Gustafsson's explicit controller :cite:p:`Gust:91`, and
+   inserts its default parameters (:math:`k_1=0.635`, :math:`k_2=-0.268`, and
+   :math:`k_3=k_4=k_5=0`).
```
```suggestion
   inserts its default parameters :math:`k_1=0.635`, :math:`k_2=-0.268`, and
   :math:`k_3=k_4=k_5=0`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: int SUNAdaptController_SetParams_ExpGus(SUNAdaptController C, sunrealtype k1, sunrealtype k2)
+
+   This user-callable function provides control over the relevant parameters
+   above, setting :math:`k_3 = k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
```
```suggestion
   This user-callable function provides control over the relevant parameters
   above for the explicit Gustafsson controller, setting :math:`k_3 = k_4 = k_5 = 0`. 
   This should be called *before* the time integrator is called to evolve the problem.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+   above, setting :math:`k_3 = k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
+
+   Note that the Gustafsson's explicit controller has the form
```
Use `.. note::`

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+   and :math:`k_2 = -\hat{k}_2`.
+
+   :param C: the SUNAdaptController_Soderlind object.
+   :param k1: parameter used within the explicit Gustafsson controller time step estimate.
+   :param k2: parameter used within the explicit Gustafsson controller time step estimate.
```
Maybe rename the input parameters here and above to `k1_hat` and `k2_hat` for additional clarity.

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
```
```suggestion
   This constructor creates and allocates memory for a SUNAdaptController_Soderlind
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+   This constructor function creates and allocates memory for a SUNAdaptController_Soderlind
+   object, set up to replicate Gustafsson's implicit controller :cite:p:`Gust:94`, and
+   inserts its default parameters (:math:`k_1=1.93`, :math:`k_2=-0.95`, :math:`k_4=1`, and
+   :math:`k_3=k_5=0`).
```
```suggestion
   inserts its default parameters :math:`k_1=1.93`, :math:`k_2=-0.95`, :math:`k_4=1`, and
   :math:`k_3=k_5=0`.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. c:function:: int SUNAdaptController_SetParams_ImpGus(SUNAdaptController C, sunrealtype k1, sunrealtype k2)
+
+   This user-callable function provides control over the relevant parameters
+   above, setting :math:`k_3 = k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
```
```suggestion
   This user-callable function provides control over the relevant parameters
   above for the implicit Gustafsson controller, setting :math:`k_3 = k_4 = k_5 = 0`.
   This should be called *before* the time integrator is called to evolve the problem.
```

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+   above, setting :math:`k_3 = k_4 = k_5 = 0`.  This should be called *before* the
+   time integrator is called to evolve the problem.
+
+   Note that the Gustafsson's implicit controller has the form
```
Use `.. note::`

gardner48 commented on doc/shared/sunadaptcontroller/SUNAdaptController_Soderlind.rst:
```diff
@@ -0,0 +1,311 @@
+   :math:`k_2 = -\hat{k}_2`, and :math:`k_4=1`.
+
+   :param C: the SUNAdaptController_Soderlind object.
+   :param k1: parameter used within the implicit Gustafsson controller time step estimate.
+   :param k2: parameter used within the implicit Gustafsson controller time step estimate.
```
Maybe rename the input parameters here and above to `k1_hat` and `k2_hat` for additional clarity.

balos1 commented:
I am not sure how (or if it possible) to use this API to tell ARKODE precisely what `h` to use based on my own user controller (via our "user" `SUNAdaptController` or a custom `SUNAdaptController`).  It seems to me like it is not possible without setting a lot of the heuristic parameters to 1.0.  However, maybe I am missing something. @drreynolds Can you explain what I would need to do?

gardner48 commented:
> I am not sure how (or if it possible) to use this API to tell ARKODE precisely what `h` to use based on my own user controller (via our "user" `SUNAdaptController` or a custom `SUNAdaptController`). It seems to me like it is not possible without setting a lot of the heuristic parameters to 1.0. However, maybe I am missing something. @drreynolds Can you explain what I would need to do?

If we consider the controller as only suggesting step sizes based on an error estimate then this kind of change would be a level above the controller API i.e., another implementation of the internal `arkAdapt` structure. In this case implementing an approach that specifies the step size requires replacing the call to `arkCheckTemporalError` with a function that uses some other strategy for step size selection.

balos1 commented:
> If we consider the controller as only suggesting step sizes based on an error estimate then this kind of change would be a level above the controller API i.e., another implementation of the internal `arkAdapt` structure. In this case implementing an approach that specifies the step size requires replacing the call to `arkCheckTemporalError` with a function that uses some other strategy for step size selection.

I see. I was hoping this PR would basically make it so that the entire process of choosing an `h` was encapsulated in some module so that all I had to do was provide a module that spits out an `h` and ARKODE would just use it no-questions asked. As you point out, we could build a module at a level above the controller API to accomplish this. With that said I think the changes in this PR are a good step in that direction, so I am happy approving it after some of the small things are addressed.

drreynolds commented on doc/arkode/guide/source/Usage/ERKStep_c_interface/User_callable.rst:
```diff
@@ -1159,36 +1160,56 @@ the code, is provided in :numref:`ARKODE.Mathematics.Adaptivity`.
+   +-----------------------------------------------------------+--------------------------------------------+-----------+
+   | Optional input                                            | Function name                              | Default   |
+   +-----------------------------------------------------------+--------------------------------------------+-----------+
+   | Provide a :c:type:`SUNAdaptController` for ERKStep to use | :c:func:`ERKStepSetAdaptController()`      | PI        |
```
The default controller for ERKStep has been PI, and not PID.  If/when we change the defaults, we can revisit this table.

drreynolds commented on doc/arkode/guide/source/Usage/ERKStep_c_interface/User_callable.rst:
```diff
@@ -1159,36 +1160,56 @@ the code, is provided in :numref:`ARKODE.Mathematics.Adaptivity`.
+   **Return value:**
+      * *ARK_SUCCESS* if successful
+      * *ARK_MEM_NULL* if the ERKStep memory is ``NULL``
+      * *ARK_MEM_FAIL* if *C* was ``NULL`` and the PID controller could not be allocated.
```
This documentation is correct as-is, since `ERKStepSetAdaptController` calls the underlying `arkSetAdaptController`, that itself defaults to PID.  I've added a note pointing out this discrepancy, including instructions for reverting to the ERKStep default.

drreynolds commented on doc/shared/sunadaptcontroller/SUNAdaptController_Description.rst:
```diff
@@ -0,0 +1,473 @@
+        int (*reset)(SUNAdaptController C);
+        int (*setdefaults)(SUNAdaptController C);
+        int (*write)(SUNAdaptController C, FILE* fptr);
+        int (*seterrorbias)(SUNAdaptController C, sunrealtype bias);
```
The bias factor is used to help "convert" local temporal error to global temporal error.  As such, I think of it as fundamentally relating to the temporal error controller (unlike the `safety` factor, that merely shrinks a predicted step size to help ensure success of subsequent error and convergence tests).

drreynolds commented on doc/shared/sunadaptcontroller/SUNAdaptController_Description.rst:
```diff
@@ -0,0 +1,473 @@
+        int (*setdefaults)(SUNAdaptController C);
+        int (*write)(SUNAdaptController C, FILE* fptr);
+        int (*seterrorbias)(SUNAdaptController C, sunrealtype bias);
+        int (*update)(SUNAdaptController C, sunrealtype h, sunrealtype dsm);
```
It updates the controller with both `h` and `dsm`, but I agree that the name could reflect the type of controller that it updates.  This is a pretty pervasive change, so I'll save it until I have some daytime bandwidth to ensure that it is changed _everywhere_.

drreynolds commented on doc/shared/sunadaptcontroller/SUNAdaptController_Description.rst:
```diff
@@ -0,0 +1,473 @@
+        SUNAdaptController_Type (*getid)(SUNAdaptController C);
+        int (*destroy)(SUNAdaptController C);
+        int (*estimatestep)(SUNAdaptController C, sunrealtype h, int p, sunrealtype dsm, sunrealtype* hnew);
+        int (*estimatestepandorder)(SUNAdaptController C, sunrealtype h, int p, sunrealtype dsm, sunrealtype* hnew, int *pnew);
+        int (*estimatemristeps)(SUNAdaptController C, sunrealtype H, sunrealtype h, int P, int p, sunrealtype DSM, sunrealtype dsm, sunrealtype* Hnew, sunrealtype *hnew);
+        int (*estimatesteptol)(SUNAdaptController C, sunrealtype H, int P, sunrealtype tolfac, sunrealtype DSM, sunrealtype dsm, sunrealtype *Hnew, sunrealtype* tolfacnew);
```
Agreed; however, this is another relatively pervasive change (as it requires my office Linux machine to rebuild the swig interfaces), so I'll hold off on implementing it until I have some daytime bandwidth.

drreynolds commented on doc/shared/sunadaptcontroller/SUNAdaptController_Description.rst:
```diff
@@ -0,0 +1,473 @@
+        int (*write)(SUNAdaptController C, FILE* fptr);
+        int (*seterrorbias)(SUNAdaptController C, sunrealtype bias);
+        int (*update)(SUNAdaptController C, sunrealtype h, sunrealtype dsm);
+        int (*updatemrih)(SUNAdaptController C, sunrealtype H, sunrealtype h, sunrealtype DSM, sunrealtype dsm);
+        int (*updatemritol)(SUNAdaptController C, sunrealtype H, sunrealtype tolfac, sunrealtype DSM, sunrealtype dsm);
```
I'll also attend to this when I have free time in my office.

drreynolds commented:
> > If we consider the controller as only suggesting step sizes based on an error estimate then this kind of change would be a level above the controller API i.e., another implementation of the internal `arkAdapt` structure. In this case implementing an approach that specifies the step size requires replacing the call to `arkCheckTemporalError` with a function that uses some other strategy for step size selection.
> 
> I see. I was hoping this PR would basically make it so that the entire process of choosing an `h` was encapsulated in some module (or set of modules) so that all I had to do was provide a module that spits out an `h` for ARKODE to use as the step not as a suggestion (this is what we need for adaptivity in SPRKStep). As you point out, we could build a module at a level above the controller API to accomplish this. With that said I think the changes in this PR are a good step in that direction, so I am happy approving it after some of the small things are addressed.

As David said, this would need to live at a higher level within ARKODE (e.g., where `ARKodeHAdaptMem` lives, or where I'd originally implemented the `SUNTimestepHeuristics` class).  That said, a user (or even a stepper within it's own `Evolve` routine) could effectively just call `arkSetFixedStep` in a one-step mode, which will already override all adjustment of the requested step size.

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_PID(C, k1, -k2, k3);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_PI(C, k1, -k2);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_I(C, k1);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_ExpGus(C, k1, k2);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_ImpGus(C, k1, k2);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -818,6 +926,256 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
+    }
+    if (idefault != 1) {
+      retval = SUNAdaptController_SetParams_ImExGus(C, k1, k2, k3, k3);
+      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
```
```suggestion
      if (retval != SUNADAPTCONTROLLER_SUCCESS) {
      (void) SUNAdaptController_Destroy(C);
```

gardner48 commented on src/arkode/arkode_io.c:
```diff
@@ -906,11 +1288,15 @@ int arkSetErrorBias(void *arkode_mem, realtype bias)
+      arkProcessError(ark_mem, ARK_CONTROLLER_ERR, "ARKODE",
+                      "arkSetErrorBias", "SUNAdaptController_SetErrorBias failure");
+      return(ARK_CONTROLLER_ERR);
+    }
```
```suggestion
  }
```

gardner48 commented on src/arkode/arkode_user_controller.c:
```diff
@@ -0,0 +1,155 @@
+  SUNAdaptController C;
+  ARKUserControlContent content;
+
+  /* Return with failure if hadapt or arkode_mem are NULL */
+  if ((hadapt == NULL) || (arkode_mem == NULL)) { return (NULL); }
```
```suggestion
  /* Return with failure if hadapt, arkode_mem, or context are NULL */
  if ((hadapt == NULL) || (arkode_mem == NULL) || (sunctx == NULL)) { return (NULL); }
```

gardner48 reviewed (changes requested):
Finished a pass through `examples`, `include`, `scripts`, and `src/arkode`. I'll start on `src/sunadaptcontroller` later today.

drreynolds commented on include/sunadaptcontroller/sunadaptcontroller_noop.h:
I agree with @gardner48 on this, but I added it specifically at the request of @balos1.  If Cody would chime into this conversation to help resolve these opposite requests, I would appreciate it.

balos1 replied:
What is the motivation to get rid of the NoOp controller? I see very minimal downside to having it and a huge upside to user friendliness and debugging.

gardner48 replied:
My impression was that the `NoOp` controller was intended for doing fixed time stepping but that use case is handled (mostly) by the `if (ark_mem->fixedstep)` logic. So I don't see why a user would create and attach a `NoOp` controller. I agree it's not much to maintain but if there's not a clear use case for a user, I don't see a need to have it publicly accessible.

drreynolds replied:
I think that this is most critical question: do we need to expose the `NoOp` controller to the user, or just use it internally within SUNDIALS.  I cannot currently think of a scenario where a user would create one (which is why it is not in the documentation), but I believe that it is in the public interface (including F2003).  If we want it to be more publicly advertised then I can add it to the documentation; if we want to remove it from the public interface then I'd appreciate some assistance with the CMake magic; if we are fine with it as-is then that is good with me.

drreynolds commented on src/arkode/arkode_arkstep_io.c:
```diff
@@ -456,6 +478,7 @@ int ARKStepSetOptimalParams(void *arkode_mem)
   ARKodeARKStepMem step_mem;
   ARKodeHAdaptMem hadapt_mem;
   int retval;
+  long int lenrw, leniw;
```
I'd like to leave it in for now, since Steven and I hope to do a more thorough study to set a more "optimal" set of default parameters.

drreynolds commented on src/arkode/arkode_erkstep_io.c:
```diff
@@ -250,6 +270,7 @@ int ERKStepSetDefaults(void* arkode_mem)
   ARKodeMem ark_mem;
   ARKodeERKStepMem step_mem;
   int retval;
+  long int lenrw, leniw;
```
I'd like to leave it in for now, since Steven and I hope to do a more thorough study to set a more "optimal" set of default parameters.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+
+int SUNAdaptController_UpdateH_ImExGus(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  SACIMEXGUS_EP(C) = SACIMEXGUS_BIAS(C) * dsm;
```
Should this be
```suggestion
  SACIMEXGUS_EP(C) = SUNMAX(SACIMEXGUS_BIAS(C) * dsm, TINY);
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
```
```suggestion
  C = SUNAdaptController_Soderlind(sunctx);
  if (C == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PID(C,
```
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

gardner48 replied:
See comments below about checking for a `NULL` input

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
```
```suggestion
  C = SUNAdaptController_Soderlind(sunctx);
  if (C == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
```
```suggestion
  C = SUNAdaptController_Soderlind(sunctx);
  if (C == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
```
```suggestion
  C = SUNAdaptController_Soderlind(sunctx);
  if (C == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
```
```suggestion
  C = SUNAdaptController_Soderlind(sunctx);
  if (C == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PI(C,
```
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

gardner48 replied:
See comments below about checking for a `NULL` input

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_I(C, DEFAULT_I_K1);
```
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

gardner48 replied:
See comments below about checking for a `NULL` input

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ExpGus(C,
```
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

gardner48 replied:
See comments below about checking for a `NULL` input

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ImpGus(C,
```
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

gardner48 replied:
See comments below about checking for a `NULL` input

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+int SUNAdaptController_UpdateH_Soderlind(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  SODERLIND_EPP(C) = SODERLIND_EP(C);
+  SODERLIND_EP(C)  = SODERLIND_BIAS(C) * dsm;
```
Should this be
```suggestion
  SODERLIND_EP(C)  = SUNMAX(SODERLIND_BIAS(C) * dsm, TINY);
```

gardner48 reviewed (changes requested):
Finished pass through `src/sunadaptcontroller`

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+int SUNAdaptController_SetParams_ImExGus(SUNAdaptController C,
+                                         sunrealtype k1e, sunrealtype k2e,
+                                         sunrealtype k1i, sunrealtype k2i)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+
+int SUNAdaptController_EstimateStep_ImExGus(SUNAdaptController C, sunrealtype h,
+                                            int p, sunrealtype dsm, sunrealtype* hnew)
+{
```
```suggestion
{
  if (C == NULL || hnew == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+}
+
+int SUNAdaptController_SetDefaults_ImExGus(SUNAdaptController C)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+}
+
+int SUNAdaptController_Write_ImExGus(SUNAdaptController C, FILE *fptr)
+{
```
```suggestion
{
  if (C == NULL || fptr == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+}
+
+int SUNAdaptController_SetErrorBias_ImExGus(SUNAdaptController C, sunrealtype bias)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+}
+
+int SUNAdaptController_UpdateH_ImExGus(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+}
+
+int SUNAdaptController_Space_ImExGus(SUNAdaptController C, long int* lenrw, long int* leniw)
+{
```
```suggestion
{
  if (C == NULL || lenrw == NULL || leniw == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/noop/sunadaptcontroller_noop.c:
```diff
@@ -0,0 +1,52 @@
+ */
+
+SUNAdaptController SUNAdaptController_NoOp(SUNContext sunctx)
+{
+  /* Create an empty controller object */
+  SUNAdaptController C = NULL;
```
```suggestion
{
  SUNAdaptController C;
  
  if (sunctx == NULL) { return NULL; }
  
  /* Create an empty controller object */
  C = NULL;
```

drreynolds replied:
This is redundant, since `SUNAdaptController_NewEmpty` already performs this check.

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptControllerContent_Soderlind content;
```
```suggestion
  SUNAdaptControllerContent_Soderlind content;

  if (sunctx == NULL) { return NULL; }
```

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+                                           sunrealtype k1, sunrealtype k2,
+                                           sunrealtype k3, sunrealtype k4,
+                                           sunrealtype k5)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+{
+  SUNAdaptController C;
```
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+{
+  SUNAdaptController C;
```
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+{
+  SUNAdaptController C;
```
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
```
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

gardner48 replied:
It is redundant but I don't think we should rely on a call to another function for input checks.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
```
```suggestion
  SUNAdaptController C;
  if (sunctx == NULL) { return NULL; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+int SUNAdaptController_SetParams_PID(SUNAdaptController C,
+                                     sunrealtype k1, sunrealtype k2,
+                                     sunrealtype k3)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+int SUNAdaptController_SetParams_PI(SUNAdaptController C,
+                                    sunrealtype k1, sunrealtype k2)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+ */
+
+int SUNAdaptController_SetParams_I(SUNAdaptController C, sunrealtype k1)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+int SUNAdaptController_SetParams_ExpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+int SUNAdaptController_SetParams_ImpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+
+int SUNAdaptController_EstimateStep_Soderlind(SUNAdaptController C, sunrealtype h,
+                                              int p, sunrealtype dsm, sunrealtype* hnew)
+{
```
```suggestion
{
  if (C == NULL || hnew == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_Reset_Soderlind(SUNAdaptController C)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_SetDefaults_Soderlind(SUNAdaptController C)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_Write_Soderlind(SUNAdaptController C, FILE *fptr)
+{
```
```suggestion
{
  if (C == NULL || fptr == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_SetErrorBias_Soderlind(SUNAdaptController C, sunrealtype bias)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_UpdateH_Soderlind(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
```
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:
```diff
@@ -0,0 +1,393 @@
+}
+
+int SUNAdaptController_Space_Soderlind(SUNAdaptController C, long int* lenrw, long int* leniw)
+{
```
```suggestion
{
  if (C == NULL || lenrw == NULL || leniw == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

drreynolds replied:
It is impossible for `C` to equal `NULL` here, since if it were then the base class would never have called this routine, and the checks on `lenrw` and `leniw` are redundant since you added those to the base class.

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+ * Free a generic SUNAdaptController (assumes content is already empty)
+ * ----------------------------------------------------------------- */
+
+void SUNAdaptController_FreeEmpty(SUNAdaptController C)
```
This function is missing from `sundials_adaptcontroller.h`
```suggestion
void SUNAdaptController_DestroyEmpty(SUNAdaptController C)
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_Destroy(SUNAdaptController C)
+{
+  if (C == NULL) return(SUNADAPTCONTROLLER_SUCCESS);
```
I think a `NULL` input should be an error
```suggestion
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

drreynolds replied:
I see little reason to return error messages if someone accidentally frees an object twice (particularly given that none of our other objects' "destroy" or "free" functions even bother to return values).

gardner48 replied:
Agreed, I didn't mean to suggest this change (got caught up in making the same suggestions above).

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+SUNAdaptController_Type SUNAdaptController_GetType(SUNAdaptController C)
+{
+  if (C == NULL) { return SUN_ADAPTCONTROLLER_NONE; }
+  return(C->ops->gettype(C));
```
```suggestion
  if (C->ops->gettype) { return C->ops->gettype(C); }
  return SUN_ADAPTCONTROLLER_NONE;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+{
+  int ier = 0;
+  *hnew = h;   /* initialize output with identity */
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL || hnew == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

drreynolds replied:
I don't see why this check on `hnew` needs to be duplicated everywhere.  If it is here then there is no reason to do it redundantly in each implementation (as requested).

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_Reset(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_SetDefaults(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C->ops->setdefaults) { ier = C->ops->setdefaults(C); }
```
```suggestion
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
  if (C->ops->setdefaults) { ier = C->ops->setdefaults(C); }
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_Write(SUNAdaptController C, FILE* fptr)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL || fptr == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

drreynolds replied:
I don't see why this check on `fptr` needs to be duplicated everywhere. If it is here then there is no reason to do it redundantly in each implementation (as requested).

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_SetErrorBias(SUNAdaptController C, sunrealtype bias)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_UpdateH(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_Space(SUNAdaptController C, long int *lenrw, long int *leniw)
+{
+  int ier = 0;
+  *lenrw = 0;   /* initialize outputs with identity */
+  *leniw = 0;
+  if (C == NULL) { return ier; }
```
```suggestion
  if (C == NULL || lenrw == NULL || leniw == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
  *lenrw = 0;   /* initialize outputs with identity */
  *leniw = 0;
```

drreynolds replied:
I don't see why these checks on `lenrw` and `leniw` need to be duplicated everywhere. If it is here then there is no reason to do it redundantly in each implementation (as requested).

gardner48 replied:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+int SUNAdaptController_EstimateStep(SUNAdaptController C, sunrealtype h, int p,
+                                    sunrealtype dsm, sunrealtype* hnew)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_Reset(SUNAdaptController C)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_SetDefaults(SUNAdaptController C)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_Write(SUNAdaptController C, FILE* fptr)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_SetErrorBias(SUNAdaptController C, sunrealtype bias)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_UpdateH(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 commented on src/sundials/sundials_adaptcontroller.c:
```diff
@@ -0,0 +1,173 @@
+
+int SUNAdaptController_Space(SUNAdaptController C, long int *lenrw, long int *leniw)
+{
+  int ier = 0;
```
```suggestion
  int ier = SUNADAPTCONTROLLER_SUCCESS;
```

gardner48 reviewed (changes requested):
Finished a pass over all changes

gardner48 commented on src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:
```diff
@@ -0,0 +1,222 @@
+SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptControllerContent_ImExGus content;
```
It is redundant but I don't think we should rely on another function to do input checks.

drreynolds commented:
A current impasse between David and I is in redundant error checking, particularly in routines that are called at every time step.  We discussed this, and our solution is as follows:

1. Insert the redundant error checking in this PR.
2. When incorporating the upcoming error-handling PR, these redundant checks should all be `#ifdef`-ed away unless SUNDIALS is built in "debug" mode.  

I added this plan here to help us to remember step 2.

drreynolds commented:
I *believe* that all comments have been addressed, aside from an in-person discussion regarding the `NoOp` implementation.

balos1 commented:
Failing on Jenkins with

```
[  2%] Building C object src/sundials/CMakeFiles/sundials_generic_obj_static.dir/sundials_band.c.o
/var/lib/jenkins/workspace/SUNDIALS_GitHub_Test_PR-358/test/tarballs/sundials-6.6.1/src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c: In function 'SUNAdaptController_EstimateStep_ImExGus':
/var/lib/jenkins/workspace/SUNDIALS_GitHub_Test_PR-358/test/tarballs/sundials-6.6.1/src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c:129:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
   const int ord = p + 1;
   ^
/var/lib/jenkins/workspace/SUNDIALS_GitHub_Test_PR-358/test/tarballs/sundials-6.6.1/src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c: In function 'SUNAdaptController_EstimateStep_Soderlind':
/var/lib/jenkins/workspace/SUNDIALS_GitHub_Test_PR-358/test/tarballs/sundials-6.6.1/src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c:329:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
   const int ord = p + 1;
```

balos1 commented:
@gardner48 Once you approve this, I will merge it.

gardner48 commented:
> @gardner48 Once you approve this, I will merge it.

Sounds good, I'll have time to look over the PR this evening. sundials-codes/answers#22 looks good, so that can be merged and the answers commit in this PR updated.

gardner48 commented on src/arkode/arkode_arkstep_io.c:
```diff
@@ -343,6 +342,29 @@ int ARKStepGetNumRelaxSolveIters(void* arkode_mem, long int* iters)
+
+
+/*===============================================================
+  DEPRECATED ARKStep optional input/output functions
```
Move `SetErrorBias` here as well

gardner48 commented on doc/arkode/guide/source/Usage/ERKStep_c_interface/User_callable.rst:
```diff
@@ -1280,6 +1341,14 @@ the code, is provided in :numref:`ARKODE.Mathematics.Adaptivity`.
+      :c:func:`ERKStepSetAdaptController` will be called, then this routine must be called
+      *second*.
+
+   .. deprecated:: x.x.x
```
Need to deprecate in header.

gardner48 commented on src/arkode/arkode_erkstep_io.c:
```diff
@@ -234,6 +233,27 @@ int ERKStepGetNumRelaxSolveIters(void* arkode_mem, long int* iters)
 
+
+/*===============================================================
+  DEPRECATED ERKStep optional input/output functions
```
Move `SetErrorBias` here as well

drreynolds commented:
> A couple of minor updates to note a deprecated function. Also looks like the output files need to be updated after recent changes in develop. I'll update the output files from Jenkins in a sec.

I don't quite understand your final sentence -- are you planning to update the full set of answer files that are failing due to the merge in the CI, or just those that relate to the Jenkins box?  Or more simply, which answer files are you not planning to update and that I'll need to handle?

gardner48 commented:
> > A couple of minor updates to note a deprecated function. Also looks like the output files need to be updated after recent changes in develop. I'll update the output files from Jenkins in a sec.
> 
> I don't quite understand your final sentence -- are you planning to update the full set of answer files that are failing due to the merge in the CI, or just those that relate to the Jenkins box? Or more simply, which answer files are you not planning to update and that I'll need to handle?

I've updated the failing output files from Jenkins so that should pass on the next run of tests. 

I also updated the answers submodule for GitHub CI to the latest commit in the answers repo but this will likely not resolve conflicting changes and another PR in the answers repo will be needed to update those files again.

drreynolds commented:
> I've updated the failing output files from Jenkins so that should pass on the next run of tests.
> 
> I also updated the answers submodule for GitHub CI to the latest commit in the answers repo but this will likely not resolve conflicting changes and another PR in the answers repo will be needed to update those files again.

Sounds good.  I'll wait for the CI tests to all finish (including failures) and then update the answers repo as necessary.

drreynolds commented:
I cannot log into ci/gitlab/lc.llnl.gov, so I don't know why that build/test is failing.
